core-bundle/src/Resources/contao/drivers/DC_Table.php

```diff
@@ -6335,38 +6329,53 @@ protected function initRoots()
 		}
 
 		// Get the IDs of all root records (list view or parent view)
-		elseif (\is_array($GLOBALS['TL_DCA'][$table]['list']['sorting']['root'] ?? null))
+		elseif (\is_array($GLOBALS['TL_DCA'][$this->strTable]['list']['sorting']['root'] ?? null))
 		{
-			$this->updateRoot(array_unique($GLOBALS['TL_DCA'][$table]['list']['sorting']['root']));
+			$this->updateRoot(array_unique($GLOBALS['TL_DCA'][$this->strTable]['list']['sorting']['root']));
 		}
 	}
 
-	protected function updateRoot(array $root)
+	protected function updateRoot(array $root, bool $isSearch = false)
 	{
-		$this->root = $root;
-
 		$table = ($GLOBALS['TL_DCA'][$this->strTable]['list']['sorting']['mode'] ?? null) == self::MODE_TREE_EXTENDED ? $this->ptable : $this->strTable;
+		$this->root = $this->eliminateNestedPages($root, $table);
+		$this->visibleRootTrails = [];
+		$this->rootChildren = [];
 
-		// Fetch visible root trails if enabled
-		if ($GLOBALS['TL_DCA'][$table]['list']['sorting']['showRootTrails'] ?? null)
+		if ($this->treeView)
 		{
-			foreach ($this->root as $id)
+			// Fetch visible root trails if enabled
+			if ($GLOBALS['TL_DCA'][$table]['list']['sorting']['showRootTrails'] ?? null)
 			{
-				$this->visibleRootTrails = array_unique(array_merge($this->visibleRootTrails, $this->Database->getParentRecords($id, $table, true)));
+				foreach ($this->root as $id)
+				{
+					$this->visibleRootTrails[] = $this->Database->getParentRecords($id, $table, true);
+				}
+
+				$this->visibleRootTrails = array_unique(array_merge(...$this->visibleRootTrails));
 			}
-		}
 
-		// $this->root might not have a correct order here, let's make sure it's ordered by sorting but only in case there are no visible root trails (aka
-		// the array does contain only top-level ids)
-		if ($this->root && empty($this->visibleRootTrails) && $this->Database->fieldExists('sorting', $table))
-		{
-			$this->root = $this->Database->execute("SELECT id FROM $table WHERE id IN (" . implode(',', $this->root) . ") ORDER BY sorting, id")->fetchEach('id');
+			// Fetch all children of the root
+			$this->rootChildren = $this->Database->getChildRecords($this->root, $table);
+
+			if ($isSearch)
+			{
+				$parents = [];
+
+				foreach ($root as $id)
+				{
+					$parents[] = $this->Database->getParentRecords($id, $table);
+				}
+
+				$this->rootChildren = array_intersect($this->rootChildren, array_merge(...$parents));
+				$this->visibleRootTrails = array_merge($this->visibleRootTrails, array_diff($this->rootChildren, $root));
+			}
 		}
 
-		// Fetch all children of the root
-		if ($this->treeView)
+		// $this->root might not have a correct order here, let's make sure it's ordered by sorting
```

Comment:
this now only applies to non-tree modes. In a parent view, we also want to ensure the correct order. But treeView already renders the top nodes from sorted IDs, if `showRootTrail` is enabled (see around line 3783). If `showRootTrail` is not enabled, the root sorting does not have any useful effect anyways, because of the mixing of levels.